### PR TITLE
perf: improve sort()

### PR DESCRIPTION
**Pull request by galvez**

### CODE REVIEW

**Comment by galvez** [comment]
https://twitter.com/anothergalvez/status/1378531421053194242

**Comment by Ethan-Arrowood** [comment]
try your benchmark script on my recent update. also please share your benchmark script so I can see how you're comparing the two

**Comment by Ethan-Arrowood** [comment]
v8 has done a great job optimizing spread and array operations so I won't be surprised if you can beat my trivial solution

**Comment by galvez** [comment]
Here you go: 

https://gist.github.com/galvez/93160cfb327ac540872425a35d9f8654

```bash
$ time node test.js sort1
$ time node test.js sort2
```

**Comment by galvez** [comment]
> v8 has done a great job optimizing spread and array operations 
> so I won't be surprised if you can beat my trivial solution

Yep, something I've learned over the years: whenever I find myself trying to overly optimize anything involving a loop or arraays in Node, chances are v8 has already thought of it 😂 So it's always best to just benchmark.

**Comment by Ethan-Arrowood** [comment]
yeah i think yours is significantly faster since you are doing everything _in-place_

**Comment by galvez** [comment]
That too.

**Comment by galvez** [comment]
Fixed conflicts.

**Comment by Ethan-Arrowood** [comment]
ty! can you add a comment roughly explaining your algorithm? Also you might need to run `npm run lint:js -- --fix`

**Comment by galvez** [comment]
Done!

**Comment by Ethan-Arrowood** [comment]
and lastly, if you can add a test or two to make sure the algorithm works. try your best to cover edge cases, but i don't expect you to get everything of course

**Comment by galvez** [comment]
> and lastly, if you can add a test or two to make sure the algorithm works.
> try your best to cover edge cases, but i don't expect you to get everything of course

Sure, will give this a shot tomorrow morning, close to bedtime for me  💤

**Comment by galvez** [comment]
@Ethan-Arrowood figured algo is broken once I started testing it -- debugging/tinkering :)

File: src/headers.js [hunk 1]
```diff
@@ -130,47 +131,20 @@ class Headers {
...
-    this[kHeaders] = sort(this[kHeaders])
+    this[kHeaders] = sort1d(this[kHeaders])
     for (let i = 0; i < this[kHeaders].length; i += 2) {
       callback.call(thisArg, this[kHeaders][i + 1], this[kHeaders][i], this)
     }
   }
 
   * [Symbol.iterator] () {
-    this[kHeaders] = sort(this[kHeaders])
+    this[kHeaders] = sort1d(this[kHeaders])
```
Comment (Ethan-Arrowood) [inline src/headers.js:141]:
```suggestion
    sort1d(this[kHeaders])
```

remove the reassignment since the algorithm mutates in place

File: src/headers.js [hunk 2]
```diff
@@ -130,47 +131,20 @@ class Headers {
   }
 
   forEach (callback, thisArg) {
-    this[kHeaders] = sort(this[kHeaders])
+    this[kHeaders] = sort1d(this[kHeaders])
```
Comment (Ethan-Arrowood) [inline src/headers.js:134]:
```suggestion
    sort1d(this[kHeaders])
```
remove the reassignment since the algorithm mutates in place

File: src/utils.js [hunk 3]
```diff
@@ -25,7 +25,27 @@ function createUndiciRequestOptions (request, signal) {
...
 }
 
+// Sorts 1-dimensional headers array by traversing it, from the end, and
+// comparing each pair to the previous pair. If the `>` comparison for
+// the previous pair is true, remove previous pair, move it to the end,
+// and reset the counter so it can restart the traversal and not miss anything
+function sort1d (arr) {
+  let i = arr.length
+  let previous
+  while (i) {
```
Comment (Ethan-Arrowood) [inline src/utils.js:35]:
```suggestion
  while (i > 0) {
```
make this condition clearer

File: src/utils.js [hunk 4]
```diff
@@ -25,7 +25,27 @@ function createUndiciRequestOptions (request, signal) {
...
+    if (arr[i - 2] > arr[i]) {
+      previous = arr.splice(i - 2, 2)
```
Comment (Ethan-Arrowood) [inline src/utils.js:37]: Is this assignment necessary?

File: test/utils.js [hunk 5]
```diff
@@ -42,3 +43,50 @@ tap.test('createUndiciRequestOptions', t => {
     t.strictDeepEqual(found, wanted)
   })
 })
+
+tap.test('sort1d', t => {
+  t.plan(1)
```
Comment (Ethan-Arrowood) [inline test/utils.js:48]: Add another test that uses the existing `.sort` to prove they have the same output. You can even use my existing solution if you need.

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]
Looking really good! Great work so far

File: test/utils.js (hunk 5 above)
Comment (galvez) [reply test/utils.js:48]: Good idea!

File: src/utils.js (hunk 4 above)
Comment (galvez) [reply src/utils.js]: Nope, removed it.

**Comment by galvez** [comment]
All done.

**Comment by Ethan-Arrowood** [comment]
fantastic work! thank you for contributing 😄

**Review by Ethan-Arrowood — APPROVED** [review]
